File: internal_testsuites/typescript/src/testsuite/startosis_add_service_with_port_spec_test/startosis_add_service_with_port_spec.test.ts
import {createEnclave} from "../../test_helpers/enclave_setup";
import {TransportProtocol } from "kurtosis-sdk";
import log from "loglevel";

const IS_PARTITIONING_ENABLED = false
const DEFAULT_DRY_RUN = false
const EMPTY_ARGS = "{}"
const SERVICE_ID = "docker-getting-started"

const TEST_NAME = "add-service-with-port-spec1"
const PORT_WITHOUT_PROTOCOL = "port1"
const PORT_WITH_PROTOCOL = "port2"

const STARLARK_SCRIPT_WITH_PORT_SPEC_SUCCESS = `
DOCKER_GETTING_STARTED_IMAGE = "docker/getting-started:latest"
SERVICE_ID = "${SERVICE_ID}"

spec = PortSpec(number = 5000, transport_protocol = "UDP")

def run(plan):
    plan.add_service(
        service_id = SERVICE_ID, 
        config = struct(
            image = DOCKER_GETTING_STARTED_IMAGE, 
            ports = {
                "${PORT_WITHOUT_PROTOCOL}": PortSpec(number = 3333),
                "${PORT_WITH_PROTOCOL}": spec
            }
        )
    )
    plan.print("httpd has been added successfully")`

jest.setTimeout(180000)

test("Test add service with optional protocol in port spec", async () => {
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)
    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }
    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {
        const runResult =
            await enclaveContext.runStarlarkScriptBlocking(STARLARK_SCRIPT_WITH_PORT_SPEC_SUCCESS, EMPTY_ARGS, DEFAULT_DRY_RUN)
        if (runResult.isErr()) {
            log.error("Unexpected error executing Starlark script");
            throw runResult.error;
        }

        const result = await enclaveContext.getServiceContext(SERVICE_ID)
        if (result.isErr()) { throw result.error }

        const serviceCtx = result.value
        const ports = serviceCtx.getPrivatePorts()

        const portsWithProtocol = ports.get(PORT_WITH_PROTOCOL)
        expect(portsWithProtocol).toBeDefined()
        if (portsWithProtocol !== undefined) {
            expect(portsWithProtocol.number).toEqual(5000)
            expect(portsWithProtocol.transportProtocol).toEqual(TransportProtocol.UDP)
        }

        const portsWithOutProtocol = ports.get(PORT_WITHOUT_PROTOCOL)
        expect(portsWithOutProtocol).toBeDefined()
        if (portsWithOutProtocol !== undefined) {
            expect(portsWithOutProtocol.number).toEqual(3333)
            expect(portsWithOutProtocol.transportProtocol).toEqual(TransportProtocol.TCP)
        }
    } finally {
        stopEnclaveFunction()
    }

    jest.clearAllTimers()
})

File: internal_testsuites/golang/testsuite/startosis_add_service_with_port_spec_test/startosis_add_service_with_port_spec_test.go
package startosis_add_service_with_port_spec_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/dzobbe/PoTE-kurtosis/api/golang/core/lib/services"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
)

const (
	testName = "add-service-with-port-spec1"

	serviceName = "docker-getting-started-success"

	starlarkScriptWithPortSpecSuccess = `
DOCKER_GETTING_STARTED_IMAGE = "docker/getting-started:latest"
SERVICE_NAME = "` + serviceName + `"

spec = PortSpec(number = 5000, transport_protocol = "UDP", wait = None)

def run(plan):
    plan.add_service(
        name = SERVICE_NAME,
        config = ServiceConfig(
            image = DOCKER_GETTING_STARTED_IMAGE, 
            ports = {
                "port1": PortSpec(number = 3333, wait = None),
                "port2": spec,
                "port3": PortSpec(number = 1234, transport_protocol = "TCP", application_protocol = "http", wait = None),
            }
        )
    )
    plan.print("httpd has been added successfully")`
)

func TestAddServiceWithPortSpec_Success(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, _, destroyEnclaveFunc, err := test_helpers.CreateEnclave(t, ctx, testName)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer func() {
		err = destroyEnclaveFunc()
		require.NoError(t, err, "An error occurred destroying the enclave after the test finished")
	}()

	// ------------------------------------- TEST RUN ----------------------------------------------
	runResult, err := test_helpers.RunScriptWithDefaultConfig(ctx, enclaveCtx, starlarkScriptWithPortSpecSuccess)
	logrus.Infof("Test Output: %v", runResult)
	require.NoError(t, err, "Unexpected error executing starlark script")

	service, err := enclaveCtx.GetServiceContext(serviceName)
	require.NoError(t, err, "Unexpected error occurred while getting service id")
	require.NotNil(t, service, "Error occurred while fetching service with name: '%v'. This may occur if service was not created")

	ports := service.GetPrivatePorts()
	require.Equal(t, services.TransportProtocol_TCP, ports["port1"].GetTransportProtocol())
	require.Equal(t, uint16(3333), ports["port1"].GetNumber())

	require.Equal(t, services.TransportProtocol_UDP, ports["port2"].GetTransportProtocol())
	require.Equal(t, uint16(5000), ports["port2"].GetNumber())

	require.Equal(t, services.TransportProtocol_TCP, ports["port3"].GetTransportProtocol())
	require.Equal(t, uint16(1234), ports["port3"].GetNumber())
	require.Equal(t, "http", ports["port3"].GetMaybeApplicationProtocol())

}
